include/revng/Yield/Support/GraphLayout/SugiyamaStyle/InternalGraph.h

```diff
@@ -0,0 +1,290 @@
+  using PathReference = layout::Path *;
+  std::variant<PathOwner, PathReference> Path;
+
+public:
+  /// Make a real edge.
+  explicit InternalEdge(std::size_t Index) :
+    IsRouted(false),
+    IsBackwards(false),
+    Index(Index),
+    Path(std::make_unique<layout::Path>()) {}
+
+  /// Make a virtual edge pointing to a path in another edge.
+  InternalEdge(std::size_t Index, InternalEdge &Another, bool IsBackwards) :
+    IsRouted(Another.IsRouted),
+    IsBackwards(IsBackwards),
+    Index(Index),
+    Path(&Another.getPathImpl()) {}
+
+  ~InternalEdge() = default;
+  InternalEdge(const InternalEdge &) = delete;
+  InternalEdge(InternalEdge &&) = default;
+  InternalEdge &operator=(const InternalEdge &) = delete;
+  InternalEdge &operator=(InternalEdge &&) = default;
+
+  std::size_t index() const { return Index; }
+  bool isVirtual() const { return std::holds_alternative<PathOwner>(Path); }
+
+  /// Adds a point to the edge path or modifies its last point if that's
+  /// sufficient, based on their coordinates.
+  void appendPoint(const layout::Point &Point) {
+    appendPointImpl(getPathImpl(), Point);
+  }
+  void appendPoint(layout::Coordinate X, layout::Coordinate Y) {
+    return appendPoint(layout::Point(X, Y));
+  }
+
+  layout::Path &getPath() {
+    revng_assert(!isVirtual(), "`getPath` should only be used on real edges.");
+    return getPathImpl();
+  }
+
+private:
+  layout::Path &getPathImpl() {
+    auto Visitor = []<typename Type>(Type &Value) -> PathReference {
+      if constexpr (std::is_same<std::decay_t<Type>, PathOwner>::value)
+        return Value.get();
+      else if constexpr (std::is_same<std::decay_t<Type>, PathReference>::value)
+        return Value;
+      else
+        static_assert(type_always_false<Type>::value, "Unknown variant type");
+    };
+    return *std::visit(Visitor, Path);
+  }
+
+  static void appendPointImpl(layout::Path &Path, const layout::Point &Point) {
+    if (Path.size() > 1) {
+      auto &First = *std::prev(std::prev(Path.end()));
+      auto &Second = *std::prev(Path.end());
+
+      auto LHS = (Point.Y - Second.Y) * (Second.X - First.X);
+      auto RHS = (Second.Y - First.Y) * (Point.X - Second.X);
+      if (LHS == RHS)
+        Path.pop_back();
+    }
+    Path.push_back(Point);
+  }
+};
+
+} // namespace detail
+
+using InternalNode = MutableEdgeNode<detail::InternalNode,
+                                     detail::InternalEdge,
+                                     false>;
+
+namespace detail {
+
+/// The result type of the \ref make method containing the newly-built
+/// graph and the lookup table necessary for the results of the layouter to be
+/// merged back into the original graph.
+template<typename NodeRef, typename EdgeRef>
+struct InternalGraphWithLookupTable;
+
+} // namespace detail
+
+class InternalGraph : public GenericGraph<InternalNode, 16, true> {
+public:
+  using GenericGraph<InternalNode, 16, true>::GenericGraph;
```

Comment:
About the `ResultType` comments here's what I propose:
- add here `using GraphWithLookupTable = detail::InternalGraphWithLookupTable<NodeRef, EdgeRef>;`
- in the prototype of `make` drop the `auto` return type and use `GraphWithLookupTable` explicitly. It makes it much more readable
- in the body of `make`,  at the end, you can drop the `using` and the `ResultType`, and just use `GraphWithLookupTable`.

Reply:
2. I'd gladly, BUT the `InternalGraphWithLookupTable` requires non-trivial template arguments that are deduced from `make`'s template parameters, so that non-`auto` return type would have to look something like `detail::InternalGraphWithLookupTable<typename llvm::GraphTraits<GraphType>::NodeRef, typename llvm::GraphTraits<GraphType>::EdgeRef>` with no place to alias those names (unless we do it inside the `template`, which I'm always opposed - it opens the possibility for somebody to explicitly override the aliases, which leads to FUD, which leads to `requires` clauses double-checking that, which mean that four letter return type (`auto`) gets bloated to four dense lines of templates.

Reply:
Uff. Ok. But please at least change `ResultType` to `GraphWithLookupTable`. It's much more meaningful.

Reply:
Renamed it.